// md-editor-v3@6.5.3/lib/es/chunks/index3.mjs
import { defineComponent as z, inject as q, ref as D, watch as J, onMounted as K, createVNode as N, reactive as X, shallowRef as G, provide as V, computed as Y, onBeforeUnmount as Z, Fragment as ee } from "vue";
import { p as O } from "./config.mjs";
import { b as $, f as j, G as U, P as te, S as le } from "./event-bus.mjs";
import { g as ne } from "./index5.mjs";
import { createSmoothScroll as oe, debounce as ie } from "@vavt/util";
const ce = `.${O}-preview > [data-line]`, w = (e, l) => +getComputedStyle(e).getPropertyValue(l).replace("px", ""), ve = (e, l) => {
  const b = ie(() => {
    e.removeEventListener("scroll", o), e.addEventListener("scroll", o), l.removeEventListener("scroll", o), l.addEventListener("scroll", o);
  }, 50), o = (E) => {
    const m = e.clientHeight, y = l.clientHeight, f = e.scrollHeight, s = l.scrollHeight, r = (f - m) / (s - y);
    E.target === e ? (l.removeEventListener("scroll", o), l.scrollTo({
      top: e.scrollTop / r
      // behavior: 'smooth'
    }), b()) : (e.removeEventListener("scroll", o), e.scrollTo({
      top: l.scrollTop * r
      // behavior: 'smooth'
    }), b());
  };
  return [
    () => {
      b().finally(() => {
        e.dispatchEvent(new Event("scroll"));
      });
    },
    () => {
      e.removeEventListener("scroll", o), l.removeEventListener("scroll", o);
    }
  ];
}, he = (e, l, b) => {
  const { view: o } = b, E = oe(), m = (i) => o.lineBlockAt(o.state.doc.line(i + 1).from).top, y = (i) => o.lineBlockAt(o.state.doc.line(i + 1).from).bottom;
  let f = [], s = [], r = [];
  const L = () => {
    f = [], s = Array.from(
      l.querySelectorAll(ce)
    ), r = s.map((n) => Number(n.dataset.line));
    const i = [...r], { lines: h } = o.state.doc;
    i[0] !== 0 && i.unshift(0);
    let H = i.shift() || 0, t = i.shift() || h;
    for (let n = 0; n < h; n++)
      n === t && (H = n, t = i.shift() || h), f.push({
        start: H,
        end: t - 1
      });
  }, R = (i, h) => {
    let H = 1;
    for (let t = s.length - 1; t - 1 >= 0; t--) {
      const n = s[t], a = s[t - 1];
      if (n.offsetTop + n.offsetHeight > h && a.offsetTop < h) {
        H = Number(a.dataset.line);
        break;
      }
    }
    for (let t = f.length - 1; t >= 0; t--) {
      const n = y(f[t].end), a = m(f[t].start);
      if (n > i && a <= i) {
        H = H < f[t].start ? H : f[t].start;
        break;
      }
    }
    return H;
  };
  let x = 0, B = 0;
  const _ = () => {
    if (B !== 0)
      return !1;
    x++;
    const { scrollDOM: i, contentHeight: h } = o;
    let H = w(l, "padding-block-start");
    const t = o.lineBlockAtHeight(i.scrollTop), { number: n } = o.state.doc.lineAt(t.from), a = f[n - 1];
    if (!a)
      return !1;
    let u = 1;
    const k = l.querySelector(`[data-line="${a.start}"]`) || l.firstElementChild?.firstElementChild, d = l.querySelector(`[data-line="${a.end + 1}"]`) || l.lastElementChild?.lastElementChild, c = i.scrollHeight - i.clientHeight, I = l.scrollHeight - l.clientHeight;
    let g = m(a.start), v = y(a.end), C = k.offsetTop, S = d.offsetTop - C;
    g === 0 && (C = 0, k === d ? (H = 0, v = h - i.offsetHeight, S = I) : S = d.offsetTop), u = (i.scrollTop - g) / (v - g);
    const A = d == l.lastElementChild?.lastElementChild ? d.offsetTop + d.clientHeight : d.offsetTop;
    if (v >= c || A > I) {
      const p = R(c, I);
      g = m(p), u = (i.scrollTop - g) / (c - g);
      const F = l.querySelector(`[data-line="${p}"]`);
      g > 0 && F && (C = F.offsetTop), S = I - C + w(l, "padding-block-start");
    }
    const T = C - H + S * u;
    E(l, T, () => {
      x--;
    });
  }, P = () => {
    if (x !== 0)
      return;
    B++;
    const { scrollDOM: i } = o, h = l.scrollTop, H = l.scrollHeight, t = i.scrollHeight - i.clientHeight, n = l.scrollHeight - l.clientHeight;
    let a = l.firstElementChild?.firstElementChild, u = l.firstElementChild?.lastElementChild;
    if (r.length > 0) {
      let A = Math.ceil(
        r[r.length - 1] * (h / H)
      ), T = r.findLastIndex((p) => p <= A);
      T = T === -1 ? 0 : T, A = r[T];
      for (let p = T; p >= 0 && p < r.length; )
        if (s[p].offsetTop > h) {
          if (p - 1 >= 0) {
            p--;
            continue;
          }
          A = -1, T = -1;
          break;
        } else {
          if (p + 1 < r.length && s[p + 1].offsetTop < h) {
            p++;
            continue;
          }
          A = r[p], T = p;
          break;
        }
      switch (T) {
        case -1: {
          a = l.firstElementChild?.firstElementChild, u = s[0];
          break;
        }
        case r.length - 1: {
          a = s[T], u = l.firstElementChild?.lastElementChild;
          break;
        }
        default:
          a = s[T], u = s[T + 1 === s.length ? T : T + 1];
      }
    }
    let k = a === l.firstElementChild?.firstElementChild ? 0 : a.offsetTop - w(a, "margin-block-start"), d = u.offsetTop, c = 0;
    const { start: I, end: g } = f[Number(a.dataset.line || 0)];
    let v = m(I);
    const C = m(
      g + 1 === o.state.doc.lines ? g : g + 1
    );
    let S = 0;
    if (C > t || u.offsetTop + u.offsetHeight > n) {
      const A = R(t, n), T = l.querySelector(`[data-line="${A}"]`);
      k = T ? T.offsetTop - w(T, "margin-block-start") : k, v = m(A), c = (h - k) / (n - k), S = t - v;
    } else a === l.firstElementChild?.firstElementChild ? (a === u && (d = u.offsetTop + u.offsetHeight + w(u, "margin-block-end")), S = C, c = Math.max(h / d, 0)) : (c = Math.max(
      (h - k) / (d - k),
      0
    ), S = C - v);
    E(e, v + S * c, () => {
      B--;
    });
  }, M = (i) => {
    const { scrollDOM: h, contentHeight: H } = o, t = h.clientHeight;
    if (H <= t || l.firstElementChild.clientHeight <= l.clientHeight || o.state.doc.lines <= f[f.length - 1]?.end)
      return !1;
    i.target === e ? _() : P();
  };
  return [
    () => {
      L(), e.addEventListener("scroll", M), l.addEventListener("scroll", M), e.dispatchEvent(new Event("scroll"));
    },
    () => {
      e.removeEventListener("scroll", M), l.removeEventListener("scroll", M);
    }
  ];
}, se = {
  tocItem: {
    type: Object,
    default: () => ({})
  },
  mdHeadingId: {
    type: Function,
    default: () => {
    }
  },
  onActive: {
    type: Function,
    default: () => {
    }
  },
  onClick: {
    type: Function,
    default: () => {
    }
  },
  scrollElementOffsetTop: {
    type: Number,
    default: 0
  }
}, Q = /* @__PURE__ */ z({
  props: se,
  setup(e) {
    const l = q("scrollElementRef"), b = q("roorNodeRef"), o = D();
    J(() => e.tocItem.active, (m) => {
      m && e.onActive(e.tocItem, o.value);
    }), K(() => {
      e.tocItem.active && e.onActive(e.tocItem, o.value);
    });
    const E = (m) => {
      if (m.stopPropagation(), e.onClick(m, e.tocItem), m.defaultPrevented)
        return;
      const y = e.mdHeadingId({
        text: e.tocItem.text,
        level: e.tocItem.level,
        index: e.tocItem.index,
        currentToken: e.tocItem.currentToken,
        nextToken: e.tocItem.nextToken
      }), f = b.value.getElementById(y), s = l.value;
      if (f && s) {
        let r = f.offsetParent, L = f.offsetTop;
        if (s.contains(r))
          for (; r && s != r; )
            L += r?.offsetTop, r = r?.offsetParent;
        const R = f.previousElementSibling;
        let x = 0;
        R || (x = w(f, "margin-block-start")), s?.scrollTo({
          top: L - e.scrollElementOffsetTop - x,
          behavior: "smooth"
        });
      }
    };
    return () => N("div", {
      ref: o,
      class: [`${O}-catalog-link`, e.tocItem.active && `${O}-catalog-active`],
      onClick: E
    }, [N("span", {
      title: e.tocItem.text
    }, [e.tocItem.text]), e.tocItem.children && e.tocItem.children.length > 0 && N("div", {
      class: `${O}-catalog-wrapper`
    }, [e.tocItem.children.map((m) => N(Q, {
      mdHeadingId: e.mdHeadingId,
      key: `${e.tocItem.text}-link-${m.level}-${m.text}`,
      tocItem: m,
      onActive: e.onActive,
      onClick: e.onClick,
      scrollElementOffsetTop: e.scrollElementOffsetTop
    }, null))])]);
  }
}), re = {
  /**
   * 编辑器的Id，务必与需要绑定的编辑器Id相同
   */
  editorId: {
    type: String,
    default: void 0
  },
  class: {
    type: String,
    default: ""
  },
  mdHeadingId: {
    type: Function,
    default: ({
      text: e
    }) => e
  },
  /**
   * 指定滚动的容器，选择器需带上对应的符号，默认预览框
   * 元素必须定位！！！！！！
   *
   * 默认：#md-editor-preview-wrapper
   */
  scrollElement: {
    type: [String, Object],
    default: void 0
  },
  theme: {
    type: String,
    default: "light"
  },
  /**
   * 高亮标题相对滚动容器顶部偏移量，即距离该值时，高亮当前目录菜单项
   *
   * 默认：20px
   */
  offsetTop: {
    type: Number,
    default: 20
  },
  /**
   * 滚动区域的固定顶部高度
   *
   * 默认：0
   */
  scrollElementOffsetTop: {
    type: Number,
    default: 0
  },
  onClick: {
    type: Function,
    default: void 0
  },
  onActive: {
    type: Function,
    default: void 0
  },
  /**
   * 滚动容器是否在web component中，默认不在
   *
   * 在其中的话通过document查询不到
   */
  isScrollElementInShadow: {
    type: Boolean,
    default: !1
  },
  /**
   * 设置与哪个区域同步，默认与内容区域同步
   *
   * >= v5.3.0
   */
  syncWith: {
    type: String,
    default: "preview"
  },
  /**
   * 控制最大显示的目录层级
   */
  catalogMaxDepth: {
    type: Number,
    default: void 0
  }
}, W = /* @__PURE__ */ z({
  name: "MdCatalog",
  props: re,
  emits: ["onClick", "onActive"],
  setup(e, l) {
    const b = e.editorId, o = `#${b}-preview-wrapper`, E = X({
      list: [],
      show: !1,
      scrollElement: e.scrollElement || o
    }), m = G(), y = D(), f = D(), s = D(), r = D(), L = G(), R = D({});
    V("scrollElementRef", f), V("roorNodeRef", r);
    const x = Y(() => {
      const t = [];
      return E.list.forEach((n, a) => {
        if (e.catalogMaxDepth && n.level > e.catalogMaxDepth)
          return;
        const {
          text: u,
          level: k,
          line: d
        } = n, c = {
          level: k,
          text: u,
          line: d,
          index: a + 1,
          active: m.value === n
        };
        if (t.length === 0)
          t.push(c);
        else {
          let I = t[t.length - 1];
          if (c.level > I.level)
            for (let g = I.level + 1; g <= 6; g++) {
              const {
                children: v
              } = I;
              if (!v) {
                I.children = [c];
                break;
              }
              if (I = v[v.length - 1], c.level <= I.level) {
                v.push(c);
                break;
              }
            }
          else
            t.push(c);
        }
      }), t;
    }), B = () => {
      if (E.scrollElement instanceof HTMLElement)
        return E.scrollElement;
      let t = document;
      return (E.scrollElement === o || e.isScrollElementInShadow) && (t = y.value?.getRootNode()), t.querySelector(E.scrollElement);
    }, _ = (t) => {
      if (t.length === 0)
        return m.value = void 0, E.list = [], !1;
      const {
        activeHead: n,
        activeIndex: a
      } = t.reduce((d, c, I) => {
        let g = 0;
        if (e.syncWith === "preview") {
          const v = r.value?.getElementById(e.mdHeadingId({
            text: c.text,
            level: c.level,
            index: I + 1,
            currentToken: c.currentToken,
            nextToken: c.nextToken
          }));
          v instanceof HTMLElement && (g = ne(v, f.value));
        } else {
          const v = L.value;
          if (v) {
            const C = v.lineBlockAt(v.state.doc.line(c.line + 1).from).top, S = v.scrollDOM.scrollTop;
            g = C - S;
          }
        }
        return g < e.offsetTop && g > d.minTop ? {
          activeHead: c,
          activeIndex: I,
          minTop: g
        } : d;
      }, {
        activeHead: t[0],
        activeIndex: 0,
        minTop: Number.MIN_SAFE_INTEGER
      });
      let u = n;
      const {
        catalogMaxDepth: k
      } = e;
      if (k && u.level > k) {
        for (let d = a; d >= 0; d--) {
          const c = t[d];
          if (c.level <= k) {
            u = c;
            break;
          }
        }
        if (u.level > k) {
          const d = t.find((c) => c.level <= k);
          d && (u = d);
        }
      }
      m.value = u, E.list = t;
    }, P = (t, n) => {
      R.value.top = n.offsetTop + w(n, "padding-block-start") + "px", e.onActive?.(t, n), l.emit("onActive", t, n);
    }, M = () => {
      _(E.list);
    }, i = (t) => {
      if (s.value?.removeEventListener("scroll", M), e.syncWith === "editor")
        s.value = L.value?.scrollDOM;
      else {
        const n = B();
        f.value = n, s.value = n === document.documentElement ? document : n;
      }
      _(t), s.value?.addEventListener("scroll", M);
    }, h = (t) => {
      L.value = t;
    };
    J([() => e.syncWith, L, () => e.catalogMaxDepth], () => {
      i(E.list);
    }), K(() => {
      r.value = y.value.getRootNode(), $.on(b, {
        name: j,
        callback: i
      }), $.on(b, {
        name: U,
        callback: h
      }), $.emit(b, te), $.emit(b, le);
    }), Z(() => {
      $.remove(b, j, i), $.remove(b, U, h), s.value?.removeEventListener("scroll", M);
    });
    const H = (t, n) => {
      e.onClick?.(t, n), l.emit("onClick", t, n);
    };
    return () => N("div", {
      class: [`${O}-catalog`, e.theme === "dark" && `${O}-catalog-dark`, e.class || ""],
      ref: y
    }, [x.value.length > 0 && N(ee, null, [N("div", {
      class: `${O}-catalog-indicator`,
      style: R.value
    }, null), N("div", {
      class: `${O}-catalog-container`
    }, [x.value.map((t) => N(Q, {
      mdHeadingId: e.mdHeadingId,
      tocItem: t,
      key: `link-${t.level}-${t.text}`,
      onActive: P,
      onClick: H,
      scrollElementOffsetTop: e.scrollElementOffsetTop
    }, null))])])]);
  }
});
W.install = (e) => (e.component(W.name, W), e);
export {
  W as M,
  ve as a,
  he as s
};
